### PR TITLE
Blackboard json

### PR DESCRIPTION
**Pull request by christianhelm**
Support for blackboards defined in json. 
Can use own blackboard or inherited blackboard.
Supports key mapping.
https://www.wrike.com/open.htm?id=816946640

### CODE REVIEW

File: alica_tests/src/test/etc/plans/behaviours/TestParameterPassingBehaviour.beh [hunk 1]
```diff
@@ -10,5 +10,29 @@
   "relativeDirectory": "",
   "requiresParameters": true,
   "runtimeCondition": null,
-  "variables": []
+  "variables": [],
```
Comment (Dale-Koenig) [inline alica_tests/src/test/etc/plans/behaviours/TestParameterPassingBehaviour.beh:13]: Must manaully set relativeDirectory to `behaviours` at the moment or everything goes in the wrong place

File: alica_tests/src/test/test_inherit_blackboard.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+};
+
+TEST_F(TestInheritBlackboard, testInheritBlackboard)
+{
+    // Use inherited blackboards and check if keys are accessible
+    ae->start();
+    ae->getAlicaClock().sleep(alica::AlicaTime::milliseconds(100));
+    auto wm = dynamic_cast<alicaTests::TestWorldModel*>(ae->getWorldModel());
+    EXPECT_EQ(wm->passedParameters["masterKeyInBehavior"], 3);
+    EXPECT_EQ(wm->passedParameters["hasBehaviorKey"], 4);
```
Comment (Dale-Koenig) [inline alica_tests/src/test/test_inherit_blackboard.cpp]: I don't get how this test is for inherited blackboards

File: alica_engine/include/engine/BasicBehaviour.h [hunk 3]
```diff
@@ -61,7 +62,7 @@ class BasicBehaviour : private RunnableObject
     const VariableGrp& getVariables() const { return _behaviour->getVariables(); }
     const Variable* getVariable(const std::string& name) const { return _behaviour->getVariable(name); };
 
-    void terminate() { stop(); };
+    void terminate(RunningPlan* rp) { stop(rp); };
```
Comment (Dale-Koenig) [inline alica_engine/include/engine/BasicBehaviour.h]: What is the point of this function?  Why not call stop directly?

File: alica_engine/src/engine/blackboard/KeyMapping.cpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+#include <engine/blackboard/KeyMapping.h>
+#include <iostream>
+#include <sstream>
+
+namespace alica
+{
+
+void KeyMapping::setInput(const Blackboard* parent_bb, Blackboard* child_bb) const
+{
+    const auto lockedParentBb = LockedBlackboardRO(*parent_bb);
```
Comment (Dale-Koenig) [inline alica_engine/src/engine/blackboard/KeyMapping.cpp]: I guess we are guaranteed no thread will be locking child blackboard when we are doing this (or setOutput)?

File: alica_tests/src/test/etc/plans/TestParameterPassing.pml [hunk 5]
```diff
@@ -35,25 +35,154 @@
...
           "id": 445396005944825225,
-          "name": ""
+          "name": "",
+          "keyMapping" : 
+          {
+            "input" : 
+            [
+              {
+                "parent" : "planOutputKey",
+                "child" : "behaviorInputKey"
```
Comment (Dale-Koenig) [inline alica_tests/src/test/etc/plans/TestParameterPassing.pml]: I think it would be clearer to call these keys parent_key, child_key

File: alica_engine/include/engine/modelmanagement/factories/BlackboardFactory.h [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+#include <yaml-cpp/yaml.h>
+
+namespace alica
+{
+class Blackboard;
+class BlackboardFactory : public Factory
+{
+public:
+    static std::unique_ptr<Blackboard> create(const YAML::Node& node);
```
Comment (Dale-Koenig) [inline alica_engine/include/engine/modelmanagement/factories/BlackboardFactory.h]: Can we just return a shared_ptr here?  Returning a unique ptr then moving it into a shared ptr is kind of weird

File: alica_engine/src/engine/RunnableObject.cpp [hunk 7]
```diff
@@ -42,7 +42,37 @@ void RunnableObject::stop()
...
+                    wrappers.begin(), wrappers.end(), [this](const auto& wrapper_ptr) { return wrapper_ptr->getAbstractPlan()->getName() == _name; });
+            assert(it != wrappers.end());
+
+            int64_t wrapperId = (*it)->getId();
+
+            BasicPlan* parentPlan = rp->getParent()->getBasicPlan();
+            const auto keyMapping = parentPlan->getKeyMapping(wrapperId);
+            auto terminateCall = [this, keyMapping, parentPlan = parentPlan]() {
+                assert(_blackboard);
+                keyMapping.setOutput(parentPlan->getBlackboard().get(), _blackboard.get());
```
Comment (Dale-Koenig) [inline alica_engine/src/engine/RunnableObject.cpp]: Can we set the output mapping the same time we set the input mapping (when initializing the behavior/plan)?

File: alica_engine/src/engine/RunnableObject.cpp [hunk 8]
```diff
@@ -52,43 +82,39 @@ void RunnableObject::start(RunningPlan* rp)
...
-            }
-            _blackboard->impl().clear();
-            if (!planAttachment->setParameters(*parentPlan->getBlackboard(), *_blackboard)) {
-                std::cerr << "Setting parameters failed, supposedly as the context has already changed.  Plan will not be scheduled" << std::endl;
-                return;
-            }
-            doInit();
-        };
-        _engine->editScheduler().schedule(initCall);
+    if (rp->getParent()) {
```
Comment (Dale-Koenig) [inline alica_engine/src/engine/RunnableObject.cpp]: I guess we could reduce a level of nesting by starting with
```
if (!rp->getParent) {
...
} else if(!_inheritBlackboard) {
```

File: alica_engine/src/engine/RunnableObject.cpp [hunk 9]
```diff
@@ -52,43 +82,39 @@ void RunnableObject::start(RunningPlan* rp)
...
+                    wrappers.begin(), wrappers.end(), [this](const auto& wrapper_ptr) { return wrapper_ptr->getAbstractPlan()->getName() == _name; });
+            assert(it != wrappers.end());
+
+            int64_t wrapperId = (*it)->getId();
+
+            BasicPlan* parentPlan = rp->getParent()->getBasicPlan();
+            const auto keyMapping = parentPlan->getKeyMapping(wrapperId);
+            auto initCall = [this, keyMapping, parentPlan = parentPlan]() {
+                assert(_blackboard);
+                keyMapping.setInput(parentPlan->getBlackboard().get(), _blackboard.get());
```
Comment (Dale-Koenig) [inline alica_engine/src/engine/RunnableObject.cpp]: I think we still need to reset child blackboard here.  The child may have protected parameters that have leftover values from previous runs.

I guess `clear()` needs to be replaced with `reset()` uch that it does not remove keys, but just resets everything to their default values

File: alica_tests/src/test/test_inherit_blackboard.cpp (hunk 2 above)
Comment (christianhelm) [reply alica_tests/src/test/test_inherit_blackboard.cpp]: It checks if a key that is only available through inheritance is available.
The second check tests if a key defined in a child is not available in the parent when they are connected by inheritance. I have added comments in the code that explain that.

Comment (Dale-Koenig) [reply alica_tests/src/test/test_inherit_blackboard.cpp]: I guess what I am confused about is the relationship between blackboards and the world model.  If this is a test for blackboards, why are we checking the world model?

Comment (christianhelm) [reply alica_tests/src/test/test_inherit_blackboard.cpp]: The blackboard itself is not accessible in the test code. The behaviors and plans read the blackboard and write into the world model. Then the test can check the world model. It's a pattern I found in several other tests and used again here.

File: alica_engine/include/engine/BasicBehaviour.h (hunk 3 above)
Comment (christianhelm) [reply alica_engine/include/engine/BasicBehaviour.h]: I'm not sure myself. I just left the naming/pattern as it was.

File: alica_engine/src/engine/blackboard/KeyMapping.cpp (hunk 4 above)
Comment (christianhelm) [reply alica_engine/src/engine/blackboard/KeyMapping.cpp]: On set input: Parent might be locked, Child is guaranteed to be not locked.
On set output: Child is guaranteed to be not locked. Parent might be locked (e.g. if blackboard is used in state transition check) 
It can only lead to problems if the lock is kept forever, otherwise it will become available at some point

File: alica_engine/src/engine/RunnableObject.cpp (hunk 7 above)
Comment (christianhelm) [reply alica_engine/src/engine/RunnableObject.cpp]: The output must be copied on termination, so I don't see a another point where this could be done.

File: alica_engine/src/engine/RunnableObject.cpp (hunk 9 above)
Comment (christianhelm) [reply alica_engine/src/engine/RunnableObject.cpp]: I have changed the way blackboards are created. Now every call of a non-inherited blackboard generates a new instance. A clear/reset should not be necessary.

File: alica_engine/src/engine/RunnableObject.cpp [hunk 10]
```diff
@@ -63,33 +101,27 @@ void RunnableObject::start(RunningPlan* rp)
...
-    } else {
-
-        BasicPlan* parentPlan = rp->getParent() ? rp->getParent()->getBasicPlan() : nullptr;
-        auto initCall = [this, parentPlan = parentPlan]() {
-            // Share Blackboard with parent if we have one, or start fresh otherwise
+    } else if (_inheritBlackboard) {
+        BasicPlan* parentPlan = rp->getParent()->getBasicPlan();
+        initCall = [this, parentPlan]() {
             if (parentPlan) {
                 _blackboard = parentPlan->getBlackboard();
```
Comment (Dale-Koenig) [inline alica_engine/src/engine/RunnableObject.cpp]: Does the check for parentPlan do anything here?  case with no parent should now be handled above right?

File: alica_engine/src/engine/RunnableObject.cpp (hunk 7 above)
Comment (Dale-Koenig) [reply alica_engine/src/engine/RunnableObject.cpp]: I suppose the scheduling can only be done here, but the keyMapping lookup could have been cached from earlier to avoid going through all the wrappers etc again.  I suppose it would have to be cached from the scheduled init() call though, so maybe easier to just go through the wrappers again here.

Mostly just complaining about the duplicated code, so maybe the solution is to refactor out the code to find the appropriate keyMapping?

File: alica_engine/include/engine/BasicBehaviour.h (hunk 3 above)
Comment (abhi1287) [reply alica_engine/include/engine/BasicBehaviour.h]: looks like this isn't used anymore and can be removed

File: alica_engine/include/engine/BasicPlan.h [hunk 11]
```diff
@@ -19,27 +20,28 @@ class BasicPlan : private RunnableObject
...
     // to share code between BasicPlan and Runnable object but not expose internals to further derived classes
     using RunnableObject::getBlackboard;
+    using RunnableObject::getInheritBlackboard;
     using RunnableObject::getName;
     using RunnableObject::getPlanContext;
-    using RunnableObject::getRequiresParameters;
     using RunnableObject::getTraceContext;
     using RunnableObject::getWorldModel;
     using RunnableObject::initExecuted;
+    using RunnableObject::setBlackboardBlueprint;
```
Comment (abhi1287) [inline alica_engine/include/engine/BasicPlan.h:31]: You can get rid of all these functions by simply passing (const BlackboardBlueprint *) in BasicPlan and BasicBehaviour constructor and storing it as const pointer as well in RunnableObject.  A nullptr would signal that no blueprint exists (inherit blackboard).

File: alica_engine/src/engine/BasicPlan.cpp [hunk 12]
```diff
@@ -81,11 +81,11 @@ void BasicPlan::doTerminate()
     _execContext.store(nullptr);
 }
 
-void BasicPlan::createChildAttachments(const Plan* plan, IPlanCreator& planCreator)
+void BasicPlan::addKeyMappings(const Plan* plan)
 {
     for (const State* state : plan->getStates()) {
         for (const ConfAbstractPlanWrapper* wrapper : state->getConfAbstractPlanWrappers()) {
-            _planAttachments.emplace(wrapper->getId(), planCreator.createPlanAttachment(wrapper->getId()));
+            _keyMappings.emplace(wrapper->getId(), wrapper->getKeyMapping());
```
Comment (abhi1287) [inline alica_engine/src/engine/BasicPlan.cpp]: Since the mappings are never going to change, keyMapping doesn't need to be duplicated every time a RunnableObject is created.  You can return a const pointer / reference from wrapper->getKeyMapping()  and just pass around the reference here.

File: alica_engine/src/engine/RunnableObject.cpp [hunk 13]
```diff
@@ -42,7 +43,40 @@ void RunnableObject::stop()
...
     }
     ++_signalState;
-    _engine->editScheduler().schedule(std::bind(&RunnableObject::doTerminate, this));
+    _engine->editScheduler().schedule([this]() { doTerminate(); });
+}
+
+std::pair<BasicPlan*, KeyMapping> RunnableObject::getParentPlanAndKeyMapping(RunningPlan* rp) const
+{
+    assert(rp->getParent()->getBasicPlan());
+    BasicPlan* parentPlan = rp->getParent()->getBasicPlan();
```
Comment (abhi1287) [inline alica_engine/src/engine/RunnableObject.cpp]: Base class (RunnableObject) having a reference to derived (BasicPlan) class is bad. Better to move mapping to the base object. Behaviour won't have children and we don't need to duplicate the mappings anyways.

File: alica_engine/include/engine/RunnableObject.h [hunk 14]
```diff
@@ -85,13 +90,14 @@ class RunnableObject
...
     AlicaTime _msInterval;
-    bool _requiresParameters;
+    bool _inheritBlackboard;
     uint8_t _flags;
     std::atomic<Counter> _execState;          // Tracks the actual executate state of the behaviour by the scheduler thread
     std::atomic<Counter> _signalState;        // Tracks the signal state from the alica main engine thread i.e. tracks start() & stop() calls
     std::atomic<RunningPlan*> _signalContext; // The running plan context when start() is called
     std::atomic<RunningPlan*> _execContext;   // The running plan context under which the behaviour is executing
     int64_t _activeRunJobId;
+    BlackboardBlueprint _blackboardBlueprint;
```
Comment (abhi1287) [inline alica_engine/include/engine/RunnableObject.h]: could be a const *

File: alica_engine/src/engine/RunnableObject.cpp [hunk 15]
```diff
@@ -52,44 +86,29 @@ void RunnableObject::start(RunningPlan* rp)
...
+    if (!rp->getParent() || !rp->getParent()->getBasicPlan()) {
+        initCall = [this]() {
             if (!_blackboard) {
-                _blackboard = std::make_shared<Blackboard>();
-            }
-            _blackboard->impl().clear();
-            if (!planAttachment->setParameters(*parentPlan->getBlackboard(), *_blackboard)) {
-                std::cerr << "Setting parameters failed, supposedly as the context has already changed.  Plan will not be scheduled" << std::endl;
-                return;
+                _blackboard = std::make_shared<Blackboard>(_blackboardBlueprint);
```
Comment (abhi1287) [inline alica_engine/src/engine/RunnableObject.cpp]: This could be quite a heavy operation every time a behaviour / plan is activated. Please put a "TBD: optimize" comment here - can fix in a later PR.

File: alica_engine/include/engine/blackboard/KeyMapping.h [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+
+class Blackboard;
+
+class KeyMapping
+{
+public:
+    KeyMapping() = default;
+    virtual ~KeyMapping() = default;
+    void setInput(const Blackboard* parent_bb, Blackboard* child_bb) const;
+    void setOutput(Blackboard* parent_bb, const Blackboard* child_bb) const;
```
Comment (abhi1287) [inline alica_engine/include/engine/blackboard/KeyMapping.h]: setInput  and setOutput method could be in the RunnableObject

File: alica_engine/include/engine/model/Behaviour.h [hunk 17]
```diff
@@ -34,11 +35,12 @@ class Behaviour : public AbstractPlan
 
     int getDeferring() const { return _deferring; }
     bool isEventDriven() const { return _eventDriven; }
-    bool getRequiresParameters() const { return _requiresParameters; }
+    bool getInheritBlackboard() const { return _inheritBlackboard; }
     int getFrequency() const { return _frequency; }
     const RuntimeCondition* getRuntimeCondition() const { return _runtimeCondition; }
     const PreCondition* getPreCondition() const { return _preCondition; }
     const PostCondition* getPostCondition() const { return _postCondition; }
+    BlackboardBlueprint getBlackboardBlueprint() const { return _blackboardBlueprint; }
```
Comment (abhi1287) [inline alica_engine/include/engine/model/Behaviour.h]: return const ptr

File: alica_engine/include/engine/model/Behaviour.h [hunk 18]
```diff
@@ -63,11 +65,11 @@ class Behaviour : public AbstractPlan
      */
     int _deferring;
     /**
-     * Specifies whether this behavior requires custom parameters to be set in its Blackboard.
-     * If so, the corresponding autogenerated function will be called
-     * Otherwise, it will simply receive a copy of its parents Blackboard
+     * Specifies whether this behavior uses its parents' blackboard.
+     * If so, it will simply receive a copy of its parents Blackboard
```
Comment (abhi1287) [inline alica_engine/include/engine/model/Behaviour.h]: it will simply receive a copy of  -> reference to

File: alica_engine/include/engine/model/Plan.h [hunk 19]
```diff
@@ -54,7 +54,8 @@ class Plan : public AbstractPlan
 
     int getFrequency() const { return _frequency; }
 
-    bool getRequiresParameters() const { return _requiresParameters; }
+    bool getInheritBlackboard() const { return _inheritBlackboard; }
+    BlackboardBlueprint getBlackboardBlueprint() const { return _blackboardBlueprint; }
```
Comment (abhi1287) [inline alica_engine/include/engine/model/Plan.h]: const *

File: alica_engine/include/engine/model/Plan.h [hunk 20]
```diff
@@ -113,11 +115,11 @@ class Plan : public AbstractPlan
      */
     int _frequency;
     /**
-     * Specifies whether this plan requires custom parameters to be set in its Blackboard.
-     * If so, the corresponding autogenerated function will be called
-     * Otherwise, it will simply receive a copy of its parents Blackboard
+     * Specifies whether this plan inherits its parents Blackboard.
+     * If so, it will simply receive a copy of its parents Blackboard
```
Comment (abhi1287) [inline alica_engine/include/engine/model/Plan.h]: reference

File: alica_engine/include/engine/modelmanagement/factories/BlackboardFactory.h [hunk 21]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include "Factory.h"
+
+#include <yaml-cpp/yaml.h>
+
+namespace alica
+{
+class BlackboardBlueprint;
+class BlackboardFactory : public Factory
```
Comment (abhi1287) [inline alica_engine/include/engine/modelmanagement/factories/BlackboardFactory.h]: BlackboardBlueprintFactory ?

File: alica_engine/src/engine/RunnableObject.cpp [hunk 22]
```diff
@@ -42,7 +43,40 @@ void RunnableObject::stop()
...
+}
+
+void RunnableObject::stop(RunningPlan* rp)
+{
+    if (!isActive(_signalState.load())) {
+        return;
+    }
+    ++_signalState;
+    if (rp->getParent()) {
+        if (!_inheritBlackboard) {
```
Comment (abhi1287) [inline alica_engine/src/engine/RunnableObject.cpp]: can club these 2 if conditions into 1

File: alica_engine/src/engine/blackboard/KeyMapping.cpp (hunk 4 above)
Comment (abhi1287) [reply alica_engine/src/engine/blackboard/KeyMapping.cpp]: so child is not required to be locked in both the cases?

File: alica_engine/src/engine/blackboard/KeyMapping.cpp [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+    }
+}
+void KeyMapping::setOutput(Blackboard* parent_bb, const Blackboard* child_bb) const
+{
+    auto lockedParentBb = LockedBlackboardRW(*parent_bb);
+    const auto lockedChildBb = LockedBlackboardRO(*child_bb);
+    for (const auto& [parentKey, childKey] : outputMapping) {
+        lockedParentBb.set(parentKey, lockedChildBb.get(childKey));
+        ALICA_DEBUG_MSG("passing " << childKey << " into " << parentKey);
+    }
```
Comment (abhi1287) [inline alica_engine/src/engine/blackboard/KeyMapping.cpp]: We should have a more graceful handling of missing keys in blackboards. 
We can probably throw exception with a warning log that either of the blackboard is missing keys or can just throw a warning message and continue. 
I am inclined towards the 2nd option because stops could be triggered by interrupts before plans could put values to output variables.

File: alica_engine/src/engine/RunnableObject.cpp [hunk 24]
```diff
@@ -52,44 +86,29 @@ void RunnableObject::start(RunningPlan* rp)
...
-                _blackboard = std::make_shared<Blackboard>();
-            }
+    } else if (!_inheritBlackboard) {
+        auto [parentPlan, keyMapping] = getParentPlanAndKeyMapping(rp);
+        initCall = [this, parentPlan, keyMapping]() {
+            _blackboard = std::make_shared<Blackboard>(_blackboardBlueprint);
+            keyMapping.setInput(parentPlan->getBlackboard().get(), _blackboard.get());
+            doInit();
+        };
+    } else if (_inheritBlackboard) {
```
Comment (abhi1287) [inline alica_engine/src/engine/RunnableObject.cpp]: could just be "else". we never want to schedule empty initCall

File: alica_engine/src/engine/RunnableObject.cpp (hunk 13 above)
Comment (christianhelm) [reply alica_engine/src/engine/RunnableObject.cpp]: I understand the problem, but can you describe your suggested solution a bit more. The KeyMapping is associated with states and can't be in the RunnableObject.

Comment (christianhelm) [reply alica_engine/src/engine/RunnableObject.cpp]: I have implemented a solution that makes use of RunningPlan

File: alica_engine/src/engine/modelmanagement/factories/KeyMappingFactory.cpp [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+
+#include "engine/modelmanagement/Strings.h"
+#include <any>
+
+namespace alica
+{
+const KeyMapping* KeyMappingFactory::create(const YAML::Node& node)
+{
+
+    auto keyMapping = new KeyMapping();
```
Comment (abhi1287) [inline alica_engine/src/engine/modelmanagement/factories/KeyMappingFactory.cpp]: this should be returned and stored in ConfAbstractPlanWrapper as a unique pointer.  By returning a const, no one would delete this memory when the engine terminates and this will show up as memory leak in all static analysis tools.

File: alica_engine/src/engine/modelmanagement/factories/BlackboardBlueprintFactory.cpp [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+#include "engine/blackboard/BlackboardBlueprint.h"
+#include "engine/modelmanagement/Strings.h"
+#include <any>
+
+namespace alica
+{
+const BlackboardBlueprint* BlackboardBlueprintFactory::create(const YAML::Node& node)
+{
+
+    auto blackboard = new BlackboardBlueprint();
```
Comment (abhi1287) [inline alica_engine/src/engine/modelmanagement/factories/BlackboardBlueprintFactory.cpp]: Same comment as keyMappingFactory. Return unique_ptr.

File: alica_engine/include/engine/model/ConfAbstractPlanWrapper.h [hunk 27]
```diff
@@ -16,11 +19,13 @@ class ConfAbstractPlanWrapper : public AlicaElement
...
     void setAbstractPlan(const AbstractPlan* abstractPlan) { _abstractPlan = abstractPlan; }
     const Configuration* getConfiguration() const { return _configuration; }
+    const KeyMapping* getKeyMapping() const { return _keyMapping; }
 
 private:
     friend ConfAbstractPlanWrapperFactory;
 
     const AbstractPlan* _abstractPlan;
     const Configuration* _configuration;
+    const KeyMapping* _keyMapping;
```
Comment (abhi1287) [inline alica_engine/include/engine/model/ConfAbstractPlanWrapper.h]: initialize with nullptr

File: alica_engine/include/engine/RunningPlan.h [hunk 28]
```diff
@@ -137,6 +137,8 @@ class RunningPlan : public std::enable_shared_from_this<RunningPlan>
     Assignment& editAssignment() { return _assignment; }
     BasicBehaviour* getBasicBehaviour() const { return _basicBehaviour; }
     BasicPlan* getBasicPlan() const { return _basicPlan; }
+    std::shared_ptr<Blackboard> getBlackboard() const;
+    const KeyMapping* getKeyMapping(int64_t wrapperId) const;
```
Comment (abhi1287) [inline alica_engine/include/engine/RunningPlan.h:141]: This still isn't clean - RunningPlan uses BasicPlan (RunnableObject)  which in turn uses RunningPlan. But that dependency existed already so lets do the cleanup in a different pr.

File: alica_engine/src/engine/RunnableObject.cpp [hunk 29]
```diff
@@ -149,4 +176,36 @@ void RunnableObject::traceInit(const std::string& type)
...
+{
+    const auto lockedParentBb = LockedBlackboardRO(*parent_bb);
+    auto& childBb = _blackboard->impl(); // Child not started yet, no other user exists, dont' use lock
+    for (const auto& [parentKey, childKey] : keyMapping->getInputMapping()) {
+        try {
+            childBb.set(childKey, lockedParentBb.get(parentKey));
+            ALICA_DEBUG_MSG("passing " << parentKey << " into " << childKey);
+        } catch (std::exception& e) {
+            ALICA_WARNING_MSG("Blackboard error passing " << parentKey << " into " << childKey << ". " << e.what());
+        }
```
Comment (abhi1287) [inline alica_engine/src/engine/RunnableObject.cpp:194]: This and output would be really nice information to put onto trace. Please create a new pr with the trace change once this is merged.

File: alica_engine/src/engine/blackboard/KeyMapping.cpp [hunk 30]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace alica
+{
+
+KeyMapping::KeyMappingList KeyMapping::getInputMapping() const
+{
+    return _inputMapping;
+}
+
+KeyMapping::KeyMappingList KeyMapping::getOutputMapping() const
```
Comment (abhi1287) [inline alica_engine/src/engine/blackboard/KeyMapping.cpp]: could be const&

File: alica_engine/include/engine/RunningPlan.h (hunk 28 above)
Comment (christianhelm) [reply alica_engine/include/engine/RunningPlan.h:141]: I have created a new task for that https://www.wrike.com/open.htm?id=840847055

File: alica_engine/src/engine/RunnableObject.cpp (hunk 29 above)
Comment (christianhelm) [reply alica_engine/src/engine/RunnableObject.cpp:194]: New task https://www.wrike.com/open.htm?id=840847670

File: alica_engine/include/engine/RunnableObject.h [hunk 31]
```diff
@@ -37,11 +39,13 @@ class RunnableObject
     const std::string& getName() { return _name; };
     AlicaTime getInterval() const { return _msInterval; };
     void setInterval(int32_t msInterval) { _msInterval = AlicaTime::milliseconds(msInterval); };
-    bool getRequiresParameters() const { return _requiresParameters; }
-    void setRequiresParameters(bool requiresParameters) { _requiresParameters = requiresParameters; }
-    void stop();
+    bool getInheritBlackboard() const { return _blackboardBlueprint == nullptr; };
+    void setBlackboardBlueprint(const BlackboardBlueprint* blackboard);
+    void stop(RunningPlan* rp);
+    void stop(); // Use only when shutdown engine
```
Comment (Dale-Koenig) [inline alica_engine/include/engine/RunnableObject.h:45]: Just rename to `onShutdown` or something

File: alica_engine/include/engine/RunnableObject.h [hunk 32]
```diff
@@ -37,11 +39,13 @@ class RunnableObject
     const std::string& getName() { return _name; };
     AlicaTime getInterval() const { return _msInterval; };
     void setInterval(int32_t msInterval) { _msInterval = AlicaTime::milliseconds(msInterval); };
-    bool getRequiresParameters() const { return _requiresParameters; }
-    void setRequiresParameters(bool requiresParameters) { _requiresParameters = requiresParameters; }
-    void stop();
+    bool getInheritBlackboard() const { return _blackboardBlueprint == nullptr; };
```
Comment (Dale-Koenig) [inline alica_engine/include/engine/RunnableObject.h:42]: What about for master plan (doesn't inherit, but no blackboard blueprint)

File: alica_engine/include/engine/model/Behaviour.h [hunk 33]
```diff
@@ -80,6 +76,11 @@ class Behaviour : public AbstractPlan
      * This behaviour's postcondition
      */
     PostCondition* _postCondition;
+    /**
+     * If nullptr it will simply receive a reference to it's parents Blackboard
```
Comment (Dale-Koenig) [inline alica_engine/include/engine/model/Behaviour.h]:
```suggestion
     * If nullptr it will simply receive a reference to its parent's Blackboard
```

**Review by Dale-Koenig — APPROVED** [review]

File: alica_engine/include/engine/RunnableObject.h (hunk 32 above)
Comment (christianhelm) [reply alica_engine/include/engine/RunnableObject.h:42]: For a master plan this could return wrong results. Where the blackboard is used (RunnableObject::start, RunnableObject::stop) this is combined with a check if there is a parent.

**Review by abhi1287 — APPROVED** [review]
